### ISSUE TITLE
Cannot use attrs post init to compute a fallback value to a field

### ISSUE THREAD

**Issue by ghisvail** [nipype/pydra#640]
I am trying to compute a fallback value for an `output_basename` field based on an `input_image`.

```python
@attrs.define(kw_only=True)
class MySpec(pydra.specs.ShellSpec):
    input_image: os.PathLike = attrs.field(...)  # mandatory

    output_basename: str = attrs.field(...)      # optional with fallback
```

Puting `pydra` aside, the documented way to achieve this with `attrs` should be with the [post init](https://www.attrs.org/en/stable/init.html#post-init) feature, i.e.

```python
...
    def __attrs_post_init__(self):
        # Fallback if output basename is unset
        self.output_basename = (
            self.output_basename or pathlib.PurePath(self.input_image).stem
        )
```

However, this triggers a `AttributeError: 'MySpecInput' object has no attribute 'files_hash'` when the corresponding `SpecInfo` object is instantiated.

So far the only workaround I found is to put the fallback logic in `formatter`, but the very same logic would need to be replicated to the output spec for each field requiring `output_basename` to compute a path to an output file.

Unless I missed a specific feature of `pydra` which would allow me to express this logic with less redundancy? To me, it's unfortunate we can't use this convenient feature from `attrs`.

**Comment by yibeichan** [comment]
Hello, @djarecka and I discussed a similar problem and had some thoughts. For `output_basename`, can we set something like, if `output_basename` is not specified, we automatically attach the function name to the input_image. For example, pydra-fsl, `output_basename` for `Split(infile=test.nii.gz)` would automatically become `test_split.nii.gz` if we don't specify it in the input_fields. (hope I'm making sense here

**Comment by ghisvail** [comment]
Indeed, `FSLSplit` is a good example. Reading through `attrs` docs on [initialization](https://www.attrs.org/en/stable/init.html), it sounds like the right solution would be to use the `@field.default` decorator like:

```python
class FSLSplitSpec(pydra.specs.ShellSpec):
    """Specifications for fslsplit."""

    input_image: os.PathLike = attrs.field(
        metadata={
            "help_string": "input image",
            "mandatory": True,
            "argstr": "",
        }
    )

    output_basename: str = attrs.field(metadata={"help_string": "output basename", "argstr": ""})

    @output_basename.default
    def _output_basename_factory(self) -> str:
        return PurePath(self.input_image).name.split(".", maxsplit=1)

    direction: str = attrs.field(
        default="t",
        metadata={
            "help_string": "split direction (x, y, z or t)",
            "argstr": "-{direction}",
            "allowed_values": {"x", "y", "z", "t"},
        },
    )
```

With this declaration, tasks instantiated with an explicit `output_basename` still work fine.

```python
>>> task = FSLSlice(input_image="volume.nii", output_basename="slice")
>>> task.cmdline
'fslsplit volume.nii.gz slice -t'
```

But produce unexpected results with the default `output_basename`:

```python
>>> FSLSplit(input_image="input.nii.gz")
>>> task.cmdline
Expected:
    'fslsplit input.nii.gz input -t'
Got:
    'fslsplit input.nii.gz Factory(factory=<function FSLSplitSpec._output_basename_factory at 0x104680ea0>, takes_self=True) -t'
```

I like this approach better than `__attrs_post_init__`. I still need to figure out why the factory is not called when the commandline is generated.

**Comment by tclose** [comment]
I believe this shouldn't be an issue with the new class
